src/architecture/serialization.md

```diff
@@ -1,533 +1,193 @@
+file and structure in it:
 
-[Ed25519 digital signature](https://ed25519.cr.yp.to/) over the binary
-serialization of the message (excluding the signature bytes,
-i.e., the last 64 bytes of the serialization).
+```rust
+module_name::proto_file_name::StructNameInProtoFile
+```
 
-**Binary presentation:** Ed25519 signature (64 bytes).  
-**JSON presentation:** hex string.
+For example, the generated `Wallet` structure, included in the
+`cryptocurrency.proto` file, which resides in the [`proto`][module] module,
+will be available using `proto::cryptocurrency::Wallet`.
 
-### Example of `transactions!` Usage
+`exonum_build` also includes a function which returns the path to the `.proto`
+files that come with the `exonum` crate - `get_exonum_protobuf_files_path()`:
 
 ```rust
-const MY_SERVICE_ID: u16 = 777;
-
-transactions! {
-    Transactions {
-        const SERVICE_ID = MY_SERVICE_ID;
+let exonum_protos = get_exonum_protobuf_files_path();
+    protobuf_generate(
+        "src/proto",
+        &["src/proto", &exonum_protos],
+        "example_mod.rs",
+    );
+```
 
-        struct MessageTwoIntegers {
-            first: u64,
-            second: u64,
-        }
-    }
+After calling the `get_exonum_protobuf_files_path()` function, users can
+import protobuf descriptions from the `exonum` crate in their `.proto` files.
+For example, `import "helpers.proto;"` can be used to get access to protobuf
+types `exonum.PublicKey`, `exonum.Hash`, etc. An example of such usage can be
+found in our [cryptocurrency example service][cryptocurrency].
+
+## Additional Validation for Protobuf-Generated Structures
+
+Protobuf is a versatile and flexible tool, which presents not only
+opportunities but also certain complications for the Exonum framework. For
+example, fields in protobuf cannot be fixed-size arrays, however, fixed-size
+arrays are required in Exonum (e.g. for hashes). It is possible to implement
+additional validations using the `.rs` protobuf-generated files. However, if
+users work with protobuf-generated structures, field validation would need to
+be performed every time they are used.
+
+To have validation performed only once for the whole structure, Exonum
+provides the conversion mechanism using the `ProtobufConvert` trait. This trait
+lets users automatically map
+their structures and the structures generated from `.proto` descriptions,
+providing a mechanism for validating protobuf-generated data. The structures
+for `ProtobufConvert` should have the same fields as the structures in
+`.proto` files, but can contain additional validation.
+
+The
+[`exonum_derive`](https://github.com/exonum/exonum/tree/master/exonum_derive)
+crate provides the ability to use structures typical for Exonum with all
+the required validations. So when using these structures users only need to
+implement `#[derive(ProtobufConvert)]` for them. If required, users can
+implement the [`ProtobufConvert`][convert] trait for any additional structures
```

Comment:
I'm afraid this sentence will be quite unclear to an average reader without additional explanation.